sync-core/src/main/java/com/cloudant/sync/query/UnindexedMatcher.java

```diff
@@ -98,14 +106,221 @@ public static UnindexedMatcher matcherWithSelector(Map<String, Object> selector)
+     *
+     * @param rev The document revision to match selector to.
+     * @return document and matcher's selector matching status.
+     */
+    public boolean matches(DocumentRevision rev) {
+        return executeSelectorTree(root, rev);
+    }
+
+    @SuppressWarnings("unchecked")
+    private boolean executeSelectorTree(QueryNode node, DocumentRevision rev) {
+        if (node instanceof AndQueryNode) {
+            boolean passed = true;
+
+            AndQueryNode andNode = (AndQueryNode) node;
+
+            for (QueryNode child: andNode.children) {
+                passed = passed && executeSelectorTree(child, rev);
+            }
+
+            return passed;
+        }
+        if (node instanceof OrQueryNode) {
+            boolean passed = false;
+
+            OrQueryNode orNode = (OrQueryNode) node;
+
+            for (QueryNode child: orNode.children) {
+                passed = passed || executeSelectorTree(child, rev);
+            }
+
+            return passed;
+        } else if (node instanceof OperatorExpressionNode) {
+            Map<String, Object> expression = ((OperatorExpressionNode) node).expression;
+
+            // Here we could have:
+            //   { fieldName: { operator: value } }
+            // or
+            //   { fieldName: { $not: { operator: value } } }
+
+            // Next evaluate the result
+            String fieldName = (String) expression.keySet().toArray()[0];
+            Map<String, Object> operatorExpression;
+            operatorExpression = (Map<String, Object>) expression.get(fieldName);
+
+            String operator = (String) operatorExpression.keySet().toArray()[0];
+
+            // First work out whether we need to invert the result when done
+            boolean invertResult = operator.equals(NOT);
+            if (invertResult) {
+                operatorExpression = (Map<String, Object>) operatorExpression.get(NOT);
+                operator = (String) operatorExpression.keySet().toArray()[0];
+            }
+
+            Object expected = operatorExpression.get(operator);
+            Object actual = ValueExtractor.extractValueForFieldName(fieldName, rev);
+
+            boolean passed = false;
+            // For array actual values, the operator expression is matched
+            // if any of the array values match it. We need to be careful
+            // to invert the match status of every candidate, rather than
+            // just flipping the result at the end.
+            //
+            // This is because { "$not": { "$eq": "white_cat" } } needs
+            // to be taken as an atomic check, meaning:
+            //   "there's an item in the array that matches `!= "white_cat"`"
+            // rather than:
+            //   "not (there's an item that matches white_cat)"
+            // The latter is satisfied using the $nin operator.
+            if (actual instanceof List) {
+                boolean currentItemPassed = false;
```

Comment:
Done in 3f1b4c2.

https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-71032f1442a4b45142f92307d7418a22L236